src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -26283,6 +27592,222 @@ class Program
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class Base
+{
+    [NotNull] public virtual string? P => throw null!;
+    public virtual string? P2 => throw null!;
+}
+public class C : Base
+{
+    public override string? P => throw null!;
+    [NotNull] public override string? P2 => throw null!;
+
+    static void M(C c, Base b)
+    {
+        b.P.ToString();
+        c.P.ToString(); // 1
+
+        b.P2.ToString(); // 2
+        c.P2.ToString();
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { source, NotNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.P.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.P").WithLocation(16, 9),
+                // (18,9): warning CS8602: Dereference of a possibly null reference.
+                //         b.P2.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.P2").WithLocation(18, 9)
+                );
+        }
+
+        [Fact]
+        public void NotNull_Property_OnVirtualProperty_OnlyOverridingGetter()
+        {
+            // Consider warning on improper nullability attributes in OHI? https://github.com/dotnet/roslyn/issues/36039
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class Base
+{
+    [NotNull] public virtual string? P { get { throw null!; } set { throw null!; } }
+    public virtual string? P2 { get { throw null!; } set { throw null!; } }
+}
+public class C : Base
+{
+    public override string? P { get { throw null!; } }
+    [NotNull] public override string? P2 { get { throw null!; } }
+
+    static void M(C c, Base b)
+    {
+        b.P.ToString();
+        c.P.ToString(); // 1
+
+        b.P2.ToString(); // 2
+        c.P2.ToString();
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { source, NotNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.P.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.P").WithLocation(16, 9),
+                // (18,9): warning CS8602: Dereference of a possibly null reference.
+                //         b.P2.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.P2").WithLocation(18, 9)
+                );
+        }
+
+        [Fact]
+        public void NotNull_01_Indexer()
+        {
+            // Warn on misused nullability attributes? https://github.com/dotnet/roslyn/issues/36073
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class COpen<TOpen>
+{
+    [NotNull]public TOpen this[int i] { get => throw null!; }
+}
+public class CClass<TClass> where TClass : class
```

Comment:
>where TClass : class [](start = 28, length = 20)

Should the constraints be `where TClass : class?` ?

Reply:
That'd be better indeed

---
In reply to: [295079619](https://github.com/dotnet/roslyn/pull/36444#discussion_r295079619) [](ancestors = 295079619)